### PR TITLE
feat(timeline): add timeline plugin

### PR DESCRIPTION
**Pull request by juliemr**
This plugin gathers test timeline information from the protractor test process,
the selenium client logs (if available), and sauce labs (if available), and
presents the output visually. This improves understanding of where latency
issues are in tests. See #674

Usage:

Add the plugin to your configuration file:

``` js
exports.config = {
  plugins: [{
    path: 'node_modules/protractor/plugins/timelinePlugin.js',

    // Output json and html will go in this folder.
    outdir: 'timelines',

    // Optional - if sauceUser and sauceKey are specified, logs from
    // SauceLabs will also be parsed after test invocation.
      sauceUser: 'Jane',
      sauceKey: 'abcdefg'
    }],
  // other configuration settings
};
```

### CODE REVIEW

File: plugins/timeline/test.js [hunk 1]
```diff
@@ -0,0 +1,124 @@
+var fs = require('fs');
```
Comment (juliemr) [inline plugins/timeline/test.js]: Done (in the /scripts/test.js file)

File: plugins/timeline/index.js [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+ * to the specified folder. A JSON data file and small index.html to view
+ * it will be created. The page uses Google Charts to show the timeline.
+ *
+ * You enable this plugin in your config file:
+ *
+ *    exports.config = {
+ *      plugins: [{
+ *        path: 'node_modules/protractor/plugins/timelinePlugin.js',
+ *
+ *        // Output json and html will go in this folder.
```
Comment (juliemr) [inline plugins/timeline/index.js]: Good point. Right now the plugins don't get passed information about the location of the config file, and the way our config parser is set up filenames are made absolute when the configuration is read. This doesn't extend well to plugins - I'll add a TODO that we should rethink this, and make it relative to the current working directory for now.

File: lib/plugins.js [hunk 3]
```diff
@@ -102,17 +102,28 @@ function pluginFunFactory(funName) {
...
 Plugins.prototype.teardown = pluginFunFactory('teardown');
 
+/**
+ * Run after the test results have been processed (any values returned will
+ * be ignored), but before the process exits. Final chance for cleanup.
+ *
+ * @return {q.Promise} A promise which resolves when the plugins have all been
+ *     torn down.
+ */
+Plugins.prototype.postResults = pluginFunFactory('postResults');
```
Comment (juliemr) [inline lib/plugins.js]: done

File: plugins/timeline/index.js [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+
+// Export
+
+var timelinePlugin = new TimelinePlugin();
+
+exports.setup = timelinePlugin.setup.bind(timelinePlugin);
+exports.teardown = timelinePlugin.teardown.bind(timelinePlugin);
+exports.postResults = timelinePlugin.postResults.bind(timelinePlugin);
+exports.TimelinePlugin = TimelinePlugin;
```
Comment (juliemr) [inline plugins/timeline/index.js]: I put the e2e tests in the timeline folder and added to the specs script. It doesn't really test much except that it doesn't blow up and fail, but the unit tests test the actual parsing.

File: plugins/timeline/index.js [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  } else {
+    self.outputResults(deferred.resolve);
+  }
+  return deferred.promise;
+};
+
+
+// Export
+
+var timelinePlugin = new TimelinePlugin();
```
Comment (juliemr) [inline plugins/timeline/index.js]: Yes, each thread is separate and will get a its own copy.

File: plugins/timeline/index.js [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+  this.clientLogAvailable = false;
+  this.outdir;
+  this.sessionId;
+  this.testProcessSetTimeoutTimestamp = 0;
+}
+
+/**
+ * Parse a selenium log in array form. For example, the logs returned
+ * from the selenium standalone server are returned as arrays.
+ */
```
Comment (juliemr) [inline plugins/timeline/index.js]: done.

File: plugins/timeline/index.js [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+};
+
+TimelinePlugin.prototype.outputResults = function(done) {
+  try {
+    fs.mkdirSync(this.outdir);
+  } catch(e) {
+    if ( e.code != 'EEXIST' ) throw e;
+  }
+  var stream = fs.createReadStream(
+      path.join(__dirname, 'timeline/indextemplate.html'));
```
Comment (juliemr) [inline plugins/timeline/index.js]: Good catch. Done.

File: plugins/timeline/index.js [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+/**
+ * Outputs information about where your Protractor test is spending its time
+ * to the specified folder. A JSON data file and small index.html to view
+ * it will be created. The page uses Google Charts to show the timeline.
+ *
+ * You enable this plugin in your config file:
+ *
+ *    exports.config = {
+ *      plugins: [{
+ *        path: 'node_modules/protractor/plugins/timelinePlugin.js',
```
Comment (juliemr) [inline plugins/timeline/index.js]: fixed.

File: plugins/timeline/index.js [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+TimelinePlugin.prototype.outputResults = function(done) {
+  try {
+    fs.mkdirSync(this.outdir);
+  } catch(e) {
+    if ( e.code != 'EEXIST' ) throw e;
+  }
+  var stream = fs.createReadStream(
+      path.join(__dirname, 'timeline/indextemplate.html'));
+  var outfile = path.join(this.outdir, 'timeline.json');
+  fs.writeFileSync(outfile, JSON.stringify(this.timeline));
```
Comment (juliemr) [inline plugins/timeline/index.js]: True, but it's not like there's any other work we're blocking, so I don't see a reason to make it async.

**Comment by juliemr** [comment]
Addressed comments, and moved tests around so hopefully they're easier to read.

**Comment by sjelin** [comment]
LGTM

**Pull request closed without merging**